cmake/cuda.cmake

```diff
@@ -1,87 +1,21 @@
 enable_language(CUDA)
 
-if(MSVC)
-    # MSVC can not find CUDA automatically
-    # Use CUDA_COMPILER PATH to define the CUDA TOOLKIT ROOT DIR
-    string(REPLACE "/bin/nvcc.exe" "" CMAKE_CUDA_ROOT_DIR ${CMAKE_CUDA_COMPILER})
-    if("${CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES}" STREQUAL "")
-        set(CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES "${CMAKE_CUDA_ROOT_DIR}/include")
-    endif()
-    if("${CMAKE_CUDA_IMPLICIT_LINK_DIRECTORIES}" STREQUAL "")
-        set(CMAKE_CUDA_IMPLICIT_LINK_DIRECTORIES "${CMAKE_CUDA_ROOT_DIR}/lib/x64")
-    endif()
-endif()
+find_package(CUDAToolkit REQUIRED)
 
 include(cmake/Modules/CudaArchitectureSelector.cmake)
 
-set(CUDA_INCLUDE_DIRS ${CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES})
-
-# Detect the CUDA architecture flags and propagate to all the project
-cas_variable_cuda_architectures(GINKGO_CUDA_ARCH_FLAGS
-    ARCHITECTURES ${GINKGO_CUDA_ARCHITECTURES}
-    UNSUPPORTED "20" "21")
-
-if (CMAKE_CXX_COMPILER_ID MATCHES "PGI|NVHPC")
-    find_package(NVHPC REQUIRED
-        HINTS
-        $ENV{NVIDIA_PATH}
-        ${CMAKE_CUDA_COMPILER}/../../..
-        )
-
-    set(CUDA_RUNTIME_LIBS_DYNAMIC ${NVHPC_CUDART_LIBRARY})
-    set(CUDA_RUNTIME_LIBS_STATIC ${NVHPC_CUDART_LIBRARY_STATIC})
-    set(CUBLAS ${NVHPC_CUBLAS_LIBRARY})
-    set(CUSPARSE ${NVHPC_CUSPARSE_LIBRARY})
-    set(CURAND ${NVHPC_CURAND_LIBRARY})
-    set(CUFFT ${NVHPC_CUFFT_LIBRARY})
-else()
-    find_library(CUDA_RUNTIME_LIBS_DYNAMIC cudart
-        HINT ${CMAKE_CUDA_IMPLICIT_LINK_DIRECTORIES})
-    find_library(CUDA_RUNTIME_LIBS_STATIC cudart_static
-        HINT ${CMAKE_CUDA_IMPLICIT_LINK_DIRECTORIES})
-
-    # CUDA 10.1/10.2 put cublas, cublasLt, cudnn in /usr/lib/<arch>-linux-gnu/, but
-    # others (<= 10.0 or >= 11) put them in cuda own directory
-    # If the environment installs several cuda including 10.1/10.2, cmake will find
-    # the 10.1/10.2 .so files when searching others cuda in the default path.
-    # CMake already puts /usr/lib/<arch>-linux-gnu/ after cuda own directory in the
-    # `CMAKE_CUDA_IMPLICIT_LINK_DIRECTORIES`, so we always put NO_DEFAULT_PATH here.
-    find_library(CUBLAS cublas
-        HINT ${CMAKE_CUDA_IMPLICIT_LINK_DIRECTORIES} NO_DEFAULT_PATH)
-    find_library(CUSPARSE cusparse
-        HINT ${CMAKE_CUDA_IMPLICIT_LINK_DIRECTORIES})
-    find_library(CURAND curand
-        HINT ${CMAKE_CUDA_IMPLICIT_LINK_DIRECTORIES})
-    find_library(CUFFT cufft
-        HINT ${CMAKE_CUDA_IMPLICIT_LINK_DIRECTORIES})
+# Detect the CUDA architecture and propagate to all the project
+cas_variable_cmake_cuda_architectures(cuda_detected_archs ${GINKGO_CUDA_ARCHITECTURES})
+if(NOT CMAKE_CUDA_ARCHITECTURES)
+    set(CMAKE_CUDA_ARCHITECTURES "${cuda_detected_archs}")
 endif()
```

Comment:
why not:
```suggestion
if(NOT CMAKE_CUDA_ARCHITECTURES)
    # Detect the CUDA architecture and propagate to all the project
    cas_variable_cmake_cuda_architectures(cuda_detected_archs ${GINKGO_CUDA_ARCHITECTURES})
    set(CMAKE_CUDA_ARCHITECTURES "${cuda_detected_archs}")
endif()
```

Reply:
Right, and also this should probably be set as a CACHE variable, otherwise we will have to regenerate the architecture variable every time.

Reply:
OTOH, with a CACHE variable, we need some invalidation logic if the GINKGO_CUDA_ARCHITECTURES value changes. I think leaving it this way should be fine